core/sessions/ldapauth/ldap.go

```diff
@@ -0,0 +1,861 @@
+package ldapauth
+
+/*
+
+The LDAP authentication module forwards the credentials in the user session request
+for authentication with a configured upstream LDAP server
+
+This module relies on the two following local database tables:
+	ldap_sessions: 	Upon successful LDAP response, creates a keyed local copy of the user
+					email
+	ldap_user_api_tokens: User created API tokens, tied to the node, storing user email.
+						  Note: user can have only one API token at a time, and token
+						  expiration is enforced
+
+User session and roles are cached and revalidated with the upstream service at the interval defined in
+the local LDAP config through the Application.sessionReaper implementation in reaper.go.
+Changes to the upstream identity server will propagate through and update local tables (web sessions, API tokens)
+by either removing the entries or updating the roles. This sync happens for every auth endpoint hit, and
+via the defined sync interval. One goroutine is created to coordinate the sync timing in the New function
+
+This implementation is read only; user mutation actions such as Delete are not supported.
+
+MFA is supported via the remote LDAP server implementation. Sufficient request time out should accommodate
+for a blocking auth call while the user responds to a potential push notification callback.
+
+*/
+
+import (
+	"crypto/subtle"
+	"database/sql"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/go-ldap/ldap/v3"
+	"github.com/smartcontractkit/sqlx"
+
+	"github.com/smartcontractkit/chainlink/v2/core/auth"
+	"github.com/smartcontractkit/chainlink/v2/core/bridges"
+	"github.com/smartcontractkit/chainlink/v2/core/config"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/logger/audit"
+	"github.com/smartcontractkit/chainlink/v2/core/services/pg"
+	"github.com/smartcontractkit/chainlink/v2/core/sessions"
+	"github.com/smartcontractkit/chainlink/v2/core/utils"
+	"github.com/smartcontractkit/chainlink/v2/core/utils/mathutil"
+)
+
+const (
+	LDAPUniqueMemberAttribute = "uniqueMember"
```

Comment:
Since we're already in `package ldapauth`, callers write `ldapauth.LDAPUniqueMemberAttribute`. What about just `ldapauth.UniqueMemberAttribute`?
```suggestion
	UniqueMemberAttribute = "uniqueMember"
```

Reply:
Yes that makes sense, updated now 🙂